torch/testing/_internal/common_device_type.py

```diff
@@ -788,6 +788,72 @@ def test_wrapper(*args, **kwargs):
                     print("Failed to instantiate {0} for op {1}!".format(test_name, op.name))
                     raise ex
 
+# Decorator that generates two set of tests with layout=torch.strided and layout=specified_layout.
+# All keyword arguments are used only for the specified layout, so they have no effect on the original test.
+class enableLayout(_TestParametrizer):
+    def __init__(self, layout: torch.layout, *,
+                 dtypes: Optional[Sequence[torch.dtype]] = None,
+                 dtypes_cpu: Optional[Sequence[torch.dtype]] = None,
+                 dtypes_cuda: Optional[Sequence[torch.dtype]] = None,
+                 only_cuda: bool = False,
+                 skip_cuda_if_no_cusparse: bool = True):
+        if not isinstance(layout, torch.layout):
+            raise RuntimeError("Expected to get torch.layout as layout, but got ", layout)
+        self.layout = layout
+        self.dtypes = set(dtypes) if dtypes is not None else None
+        self.dtypes_cpu = set(dtypes_cpu) if dtypes_cpu is not None else None
+        self.dtypes_cuda = set(dtypes_cuda) if dtypes_cuda is not None else None
+        self.only_cuda = only_cuda
+        self.skip_cuda_if_no_cusparse = skip_cuda_if_no_cusparse
+
+    def _parametrize_test(self, test, generic_cls, device_cls):
+        """ Parameterizes the given test function with boolean sparse_csr flag. """
+        # if self.layout == torch.strided generate only one set of tests
+        maybe_layout = (self.layout,) if self.layout != torch.strided else tuple()
+        for layout in (torch.strided, *maybe_layout):
+            # get dtypes specified with @dtypes, @dtypesIfCPU, @dtypesIfCUDA
+            dtypes = set(device_cls._get_dtypes(test))
+
+            # replace dtypes only with non-default layout
+            non_default_layout = (layout != torch.strided)
+            if non_default_layout:
+                if self.dtypes_cpu is not None and device_cls.device_type == 'cpu':
+                    dtypes = self.dtypes_cpu
+                if self.dtypes_cuda is not None and device_cls.device_type == 'cuda':
+                    dtypes = self.dtypes_cuda
+                elif self.dtypes is not None:
+                    dtypes = self.dtypes
+
+            for dtype in dtypes:
+                layout_str = str(layout).replace('torch.', '_')
+                # Construct the test name
+                test_name = '{}{}_{}{}'.format(test.__name__,
+                                               layout_str if non_default_layout else '',
+                                               device_cls.device_type,
+                                               _dtype_test_suffix(dtype))
+
+                # Construct parameter kwargs to pass to the test
+                param_kwargs = {'layout': layout}
+                _update_param_kwargs(param_kwargs, 'dtype', dtype)
```

Comment:
Do we also need to pass `device` here or is this done by the instantiation function?

Reply:
This is done by the instantiation function.